### ISSUE TITLE
MarkersVisual: Blending not working well with depth test on

### ISSUE THREAD

**Issue by langurmonkey** [vispy/vispy#1160]
I'm getting weird results when I enable both the depth test and blending in MarkersVisual. It seems that the outline of the markers is not correctly blended. It works well, however, when I disable the depth testing. Seems like a problem with the antialiasing (works well when I set `scatter.antialias = 0` but it looks terrible), and happens with all marker symbols.

![vispy-markers-issue](https://cloud.githubusercontent.com/assets/1049893/12529704/b36bf2ec-c1c1-11e5-8902-dc7b2ef8a4d5.png)

Here is a snippet demonstrating the issue.

```python
import sys
import numpy as np
from vispy import scene
from vispy.scene import visuals

# create data cloud
pos = np.random.normal(size=(1000, 3), scale=0.2)


#
# Make a canvas and add simple view
#
canvas = scene.SceneCanvas(keys='interactive', show=True, bgcolor='w')

view = canvas.central_widget.add_view()
view.camera = 'turntable'
view.camera.fov = 60

scatter = visuals.Markers(parent=view.scene)
# scatter.antialias = 0
scatter.set_data(pos=pos, edge_color=(0.0, 0.0, 0.0, 1.0), face_color=(0.6, 0.5, 0.4, 1.0), size=15)
scatter.set_gl_state(depth_test=True, blend=True, blend_func=('src_alpha', 'one_minus_src_alpha'))

# Add axes
axis = visuals.XYZAxis(parent=view.scene)

if __name__ == '__main__' and sys.flags.interactive == 0:
    canvas.app.run()
```

**Comment by rougier** [comment]
Yes, this is one of the limit of the method. The fix would be to sort all points such as to display them from back to front but this would just kill performances.

**Comment by langurmonkey** [comment]
Oh, that's a shame. Why is it off limits? Is it only because of AA? Is there a way to display a 3D scatter plot with the depth test and blending working in vispy (I need the fog effect)?

**Comment by almarklein** [comment]
@rougier did you not have order independent transparency working? Or was that only in a proof of concept?

**Comment by rougier** [comment]
Yes, good point. But I never tried with markers. Not sure it'll solve the problem.

**Comment by astrofrog** [comment]
I'm having this issue too. Is all the code for this in ``vispy/visuals/markers.py``? If so, I'll take a look at some point to see if we can rethink the algorithm to get this to work properly. I think it's important to get right because otherwise 3D visualization of markers with different colors is very confusing (since they don't pass in front of or behind other markers in an intuitive way).

**Comment by rougier** [comment]
If you want to have markers without antialias, it's fine. Problems arrive when you mixed antialias and depth testing. This would require to sort the markers according to depth in order to display them from back to front. Else, the first marker begin draw consider th background white and the antialias is made against this white background (this why some markers get the white border)

**Comment by astrofrog** [comment]
@rougier - I'm still new to OpenGL, but just to make sure I understand, the depth testing is built in to OpenGL itself, correct? That is, VisPy just draws markers at 3D coordinates and then the shader says what color each part of the marker should be? Is that right, or does VisPy have to handle some of the depth testing? Is it not possible to return semi-transparent colors instead of blending the anti-aliased region with the background color?

**Comment by astrofrog** [comment]
Another question - is there any way to have OpenGL handle the anti-aliasing by using some kind of multi/super sampling of each final pixel?

**Comment by rougier** [comment]
To your first question, yes OpenGL handles depth tests unless you instruct it to not do it (GL_DEPTH_TEST on/off). But the "problem" is that it is all internal such that when we draw the antialias border, we don't know what has been already drawn or not. A potential solution would be to first draw marker without antialias and then draw the antialias border but this would slow things a lot. You can also use non antialias markers using transparent colors. If you have a lot of markers, the output quality should be ok.

To your second question, yes there are some supersampling method to alias the whole scene but results are not really good for small details such as lines and markers. Almar did some  experiments at:

https://github.com/vispy/vispy/wiki/Tech.%20Antialiasing

Also, another limit of the marked techniques is that all fragments (= pixels) of a marker have the same depth (in the marked demo). If you use big marker, you will sees some "flickering" when a whole marker is draw front then back (for example just use 2 big markers and you should see the effect).
